File: android/src/main/java/com/unact/yandexmapkit/YandexPedestrian.java
package com.unact.yandexmapkit;

import android.content.Context;

import androidx.annotation.NonNull;

import com.yandex.mapkit.RequestPoint;
import com.yandex.mapkit.transport.TransportFactory;
import com.yandex.mapkit.transport.masstransit.PedestrianRouter;
import com.yandex.mapkit.transport.masstransit.Session;
import com.yandex.mapkit.transport.masstransit.TimeOptions;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;

public class YandexPedestrian implements MethodCallHandler {
    private final PedestrianRouter pedestrianRouter;
    private final BinaryMessenger binaryMessenger;
    @SuppressWarnings({"MismatchedQueryAndUpdateOfCollection"})
    private final Map<Integer, YandexPedestrianSession> pedestrianSessions = new HashMap<>();

    public YandexPedestrian(Context context, BinaryMessenger messenger) {
        TransportFactory.initialize(context);

        pedestrianRouter = TransportFactory.getInstance().createPedestrianRouter();
        binaryMessenger = messenger;
    }

    @Override
    @SuppressWarnings({"SwitchStatementWithTooFewBranches"})
    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        switch (call.method) {
            case "requestRoutes":
                requestRoutes(call, result);
                break;
            default:
                result.notImplemented();
                break;
        }
    }

    // public Session requestRoutes
    // (Point source, Point destination, TimeOptions timeOptions, RouteListener routeListener)
    @SuppressWarnings({"unchecked", "ConstantConditions"})
    private void requestRoutes(final MethodCall call, final MethodChannel.Result result) {
        Map<String, Object> params = (Map<String, Object>) call.arguments;
        Integer sessionId = (Integer) params.get("sessionId");

        List<RequestPoint> points = new ArrayList<>();
        for (Map<String, Object> pointParams : (List<Map<String, Object>>) params.get("points")) {
            points.add(Utils.requestPointFromJson(pointParams));
        }

        Session session = pedestrianRouter.requestRoutes(
                points,
                //(TimeOptions) params.get("timeOptions"),
                new TimeOptions(),
                new YandexPedestrianListener(result)
        );

        YandexPedestrianSession pedestrianSession = new YandexPedestrianSession(
                sessionId,
                session,
                binaryMessenger,
                new YandexPedestrian.PedestrianCloseListener()
        );

        pedestrianSessions.put(sessionId, pedestrianSession);
    }

    public class PedestrianCloseListener {
        public void onClose(int id) {
            pedestrianSessions.remove(id);
        }
    }
}

File: ios/Classes/YandexPedestrian.swift
import Flutter
import UIKit
import YandexMapsMobile

public class YandexPedestrian: NSObject, FlutterPlugin {
  private let methodChannel: FlutterMethodChannel!
  private let pluginRegistrar: FlutterPluginRegistrar!
  private let pedestrianRouter: YMKPedestrianRouter!
  private var pedestrianSessions: [Int: YandexPedestrianSession] = [:]

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "yandex_mapkit/yandex_pedestrian",
      binaryMessenger: registrar.messenger()
    )

    let plugin = YandexPedestrian(channel: channel, registrar: registrar)

    registrar.addMethodCallDelegate(plugin, channel: channel)
  }

  public required init(channel: FlutterMethodChannel, registrar: FlutterPluginRegistrar) {
    self.pluginRegistrar = registrar
    self.methodChannel = channel
      self.pedestrianRouter = YMKTransport.sharedInstance().createPedestrianRouter()

    super.init()

    self.methodChannel.setMethodCallHandler(self.handle)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "requestRoutes":
      requestRoutes(call, result)
      break
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func requestRoutes(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    let params = call.arguments as! [String: Any]
    let sessionId = params["sessionId"] as! Int
    let requestPoints = (params["points"] as! [[String: Any]]).map {
      (pointParams) -> YMKRequestPoint in Utils.requestPointFromJson(pointParams)
    }
      let session = pedestrianRouter.requestRoutes(
    with: requestPoints, timeOptions:YMKTimeOptions(),
    routeHandler: {(pedestrianResponse: [YMKMasstransitRoute]?, error: Error?) -> Void in
        self.pedestrianSessions[sessionId]?.handleResponse(pedestrianResponse: pedestrianResponse, error: error, result: result)
      }
    )

      pedestrianSessions[sessionId] = YandexPedestrianSession(
      id: sessionId,
      session: session,
      registrar: pluginRegistrar,
      onClose: { (id) in self.pedestrianSessions.removeValue(forKey: id) }
    )
  }
}
